lib/src/rocket.rs

```diff
@@ -191,20 +191,49 @@ impl Rocket {
         }
     }
 
-    #[inline]
+    #[inline(always)]
     pub(crate) fn dispatch<'s, 'r>(
         &'s self,
         request: &'r mut Request<'s>,
-        data: Data,
+        data: Data
     ) -> Response<'r> {
         info!("{}:", request);
 
         // Do a bit of preprocessing before routing; run the attached fairings.
         self.preprocess_request(request, &data);
+
+        // Run the request fairings.
         self.fairings.handle_request(request, &data);
 
+        // Check if the request is a `HEAD` request.
+        let was_head_request = request.method() == Method::Head;
+
+        // Route the request and run the user's handlers.
+        let mut response = self.route_and_process(request, data);
+
+
+        // Add the 'rocket' server header to the response and run fairings.
+        // TODO: If removing Hyper, write out `Date` header too.
+        response.set_header(Header::new("Server", "Rocket"));
+
+        self.fairings.handle_response(request, &mut response);
+
+        // Strip the body if this is a `HEAD` request.
+        if was_head_request {
+            response.strip_body();
+        }
+
+        response
+    }
+
+
+    fn route_and_process<'s, 'r>(
+        &'s self,
+        request: &'r Request<'s>,
```

Comment:
This should be `&'r mut` instead of just `&'r`, since this function might modify `request`

Reply:
Absolutely! Great catch! This mistake leads to highly undefined behavior when the `&mut Request` is created later from it.